File: afcm-backend/app/schemas/food.py
from pydantic import BaseModel


class OtherNutrient(BaseModel):
    name: str
    amount: float
    unit: str


class Nutrition(BaseModel):
    calories: int
    protein_g: float
    carbs_g: float
    fat_g: float
    fiber_g: float
    other_nutrients: list[OtherNutrient]


class FoodItem(BaseModel):
    food_id: str
    name: str
    brand: str | None = None
    barcode: str | None = None
    confidence: float | None = None
    reference_weight_g: float
    serving_size_label: str | None = None
    nutrition: Nutrition


class FoodAnalysisResponse(BaseModel):
    food: FoodItem


class BarcodeRequest(BaseModel):
    barcode: str


class BarcodeResponse(BaseModel):
    food: FoodItem


class CalculateWeightRequest(BaseModel):
    """Send food_id for a food the AI or barcode lookup returned, or food_name for manual entry."""

    food_id: str | None = None
    food_name: str | None = None
    weight_g: float


class CalculateWeightResponse(BaseModel):
    food_id: str
    food_name: str
    weight_g: float
    nutrition: Nutrition

File: frontend/src/types/food.ts
export interface OtherNutrient {
  name: string;
  amount: number;
  unit: string;
}

export interface Nutrition {
  calories: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  fiber_g: number;
  other_nutrients: OtherNutrient[];
}

export type FoodSource = 'camera' | 'upload' | 'barcode' | 'manual';

/** A recognised or looked-up food. `nutrition` is for `reference_weight_g`. */
export interface FoodItem {
  food_id: string;
  name: string;
  brand?: string | null;
  barcode?: string | null;
  /** 0 to 1. Only present for AI image recognition. */
  confidence?: number | null;
  reference_weight_g: number;
  /** e.g. "1 bar (40 g)" for packaged products */
  serving_size_label?: string | null;
  nutrition: Nutrition;
}

/** POST /food/analyze (multipart/form-data, field "image") */
export interface FoodAnalysisResponse {
  food: FoodItem;
}

/** POST /food/barcode */
export interface BarcodeRequest {
  barcode: string;
}
export interface BarcodeResponse {
  food: FoodItem;
}

/**
 * POST /food/calculate-weight
 * Send `food_id` for a food the AI or barcode lookup returned, or `food_name` for manual entry.
 */
export interface CalculateWeightRequest {
  food_id?: string;
  food_name?: string;
  weight_g: number;
}
export interface CalculateWeightResponse {
  food_id: string;
  food_name: string;
  weight_g: number;
  nutrition: Nutrition;
}

export type FoodInputMode = 'camera' | 'upload' | 'barcode';
